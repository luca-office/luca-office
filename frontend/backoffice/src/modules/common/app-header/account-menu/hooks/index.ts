export * from "./use-account-menu"
