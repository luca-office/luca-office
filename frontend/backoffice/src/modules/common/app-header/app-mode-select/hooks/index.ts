export * from "./use-app-mode-select"
