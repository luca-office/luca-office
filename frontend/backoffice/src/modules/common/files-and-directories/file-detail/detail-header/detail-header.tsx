import {css} from "@emotion/react"
import * as React from "react"
import {CardHeader, DeleteOrArchiveEntityButton, Heading, Icon} from "shared/components"
import {HeadingLevel, IconName} from "shared/enums"
import {DeleteEntityHook} from "shared/models"
import {Flex, spacing, spacingMedium, spacingSmall, textEllipsis} from "shared/styles"

export interface FileDetailHeaderProps {
  readonly icon: IconName
  readonly title: string
  readonly entityId: UUID
  readonly deleteHook?: (id: string) => DeleteEntityHook
  readonly onDelete?: () => void
  readonly disabled: boolean
}

export const DetailHeader: React.FC<FileDetailHeaderProps> = ({
  icon,
  title,
  entityId,
  deleteHook,
  onDelete,
  disabled
}) => {
  return (
    <CardHeader customStyles={styles.header} hasShadow hasGreyBackground>
      <div css={styles.title}>
        <Icon customStyles={styles.icon} name={icon} />
        <Heading customStyles={textEllipsis} level={HeadingLevel.h3}>
          {title}
        </Heading>
      </div>
      <div css={[Flex.row, {flexShrink: 0}]}>
        {deleteHook && (
          <DeleteOrArchiveEntityButton
            disabled={disabled}
            entityId={entityId}
            useDeleteHook={deleteHook}
            onSuccess={onDelete}
          />
        )}
      </div>
    </CardHeader>
  )
}

const styles = {
  header: css({
    display: "flex",
    justifyContent: "space-between"
  }),
  title: css({
    display: "flex",
    alignItems: "center",
    minWidth: 0
  }),
  viewerButton: css({
    marginRight: spacingSmall,
    flexShrink: 0,
    width: "auto",
    padding: spacing(0, spacingMedium)
  }),
  icon: css({
    marginRight: spacingSmall
  })
}
