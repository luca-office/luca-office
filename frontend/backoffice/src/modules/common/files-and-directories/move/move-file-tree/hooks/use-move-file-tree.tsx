import {useState} from "react"

export interface UseMoveFileTreeHook {
  readonly expandedDirectories: string[]
  readonly toggleExpandedDirectory: (id: UUID) => void
}

export const useMoveFileTree = (): UseMoveFileTreeHook => {
  const [expandedDirectories, setExpandedDirectories] = useState<string[]>([])

  const toggleExpandedDirectory = (id: UUID) =>
    setExpandedDirectories(
      expandedDirectories.includes(id) ? expandedDirectories.filter(id => id !== id) : expandedDirectories.concat(id)
    )

  return {
    expandedDirectories,
    toggleExpandedDirectory
  }
}
