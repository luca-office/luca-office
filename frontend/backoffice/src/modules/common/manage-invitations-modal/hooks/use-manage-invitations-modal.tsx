import {trim} from "lodash-es"
import * as React from "react"
import {emailRegexPattern} from "shared/utils"
import {EmailAddressesMetadata} from "../models/email-addresses-metadata"
import {getEmailAddressesMetadata} from "../utils/invite-attendees-modal-utils"

export interface UseManageInvitationsModalHook {
  readonly alreadyInvitedAddresses: string[]
  readonly createInvitations: () => void
  readonly createInvitationsLoading: boolean
  readonly emailAddressesMetadata: EmailAddressesMetadata
  readonly emailAddressesValue: string
  readonly onDeleteEmailAddress: (address: string) => void
  readonly setEmailAddressesValue: (addresses: string) => void
}

export interface UseManageInvitationsModalProps {
  readonly existingInvitations: string[]
  readonly handleInvitations: (uniqueAddresses: string[]) => void | Promise<void>
  readonly invitationProcessLoading: boolean
  readonly onEmailAddressesChange?: (emailAddresses: string[]) => void
}

export const useManageInvitationsModal = ({
  existingInvitations,
  invitationProcessLoading,
  handleInvitations,
  onEmailAddressesChange
}: UseManageInvitationsModalProps): UseManageInvitationsModalHook => {
  const isMounted = React.useRef<boolean>(false)

  const [emailAddressesValue, setEmailAddressesValue] = React.useState("")

  const emailAddressesMetadata = getEmailAddressesMetadata(emailAddressesValue, existingInvitations)

  const handleEmailAddressesChange = (addresses: string) => {
    if (onEmailAddressesChange === undefined) {
      return
    }

    const emails = addresses
      .split(";")
      .map(address => trim(address))
      .filter(addresses => !!addresses && emailRegexPattern.test(addresses))
    onEmailAddressesChange(emails)
  }

  const onDeleteEmailAddress = (address: string) => {
    const allAddresses = [
      ...emailAddressesMetadata.uniqueAddresses,
      ...emailAddressesMetadata.duplicatedAddresses,
      ...emailAddressesMetadata.alreadyInvitedAddresses
    ]
    const targetIndex = allAddresses.lastIndexOf(address)

    const addresses = [...allAddresses]
    addresses.splice(targetIndex, 1)

    setEmailAddressesValue(addresses.join("; "))
    handleEmailAddressesChange(emailAddressesValue.replace(address, ""))
  }
  const onSubmitInvitation = () => {
    const result = handleInvitations(emailAddressesMetadata.uniqueAddresses)
    if (result instanceof Promise) {
      result.then(() => isMounted.current && setEmailAddressesValue(""))
    }
  }

  const handleSetEmailAddressesValue = (addresses: string) => {
    setEmailAddressesValue(addresses)
    handleEmailAddressesChange(addresses)
  }

  React.useEffect(() => {
    isMounted.current = true
    return () => {
      isMounted.current = false
    }
  }, [])

  return {
    alreadyInvitedAddresses: existingInvitations,
    createInvitations: onSubmitInvitation,
    createInvitationsLoading: invitationProcessLoading,
    emailAddressesMetadata,
    emailAddressesValue,
    onDeleteEmailAddress,
    setEmailAddressesValue: handleSetEmailAddressesValue
  }
}
