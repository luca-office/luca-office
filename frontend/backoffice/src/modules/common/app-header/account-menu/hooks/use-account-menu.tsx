import i18next from "i18next"
import {useCallback, useState} from "react"
import {useDispatch} from "react-redux"
import {useCheckLogin} from "shared/graphql/hooks"
import {LucaAppLanguage, UserAccountForLogin} from "shared/models"
import {Option} from "shared/utils"
import {useLogout} from "../../../../../graphql/hooks"
import {navigateToRouteAction} from "../../../../../redux/actions/navigation-action"
import {Route} from "../../../../../routes"

export interface UseAccountMenuHook {
  readonly isMenuVisible: boolean
  readonly activeLanguage: string
  readonly logout: () => void
  readonly openAccountSettings: () => void
  readonly toggleMenu: () => void
  readonly userAccount: Option<UserAccountForLogin>
  readonly setStoredSelectedLanguage: (language: LucaAppLanguage) => void
}

export const useAccountMenu = (): UseAccountMenuHook => {
  const [isMenuVisible, toggleMenuVisibility] = useState(false)
  const {logout} = useLogout()
  const {account: userAccount} = useCheckLogin()
  const dispatch = useDispatch()

  const openAccountSettings = useCallback(() => {
    dispatch(navigateToRouteAction(Route.MyAccount))
  }, [])

  const toggleMenu = () => {
    toggleMenuVisibility(!isMenuVisible)
  }

  const handleSetStoredSelectedLanguage = (language: LucaAppLanguage) => i18next.changeLanguage(language)

  const activeLanguage = i18next.language

  return {
    isMenuVisible,
    activeLanguage,
    logout,
    openAccountSettings,
    toggleMenu,
    userAccount,
    setStoredSelectedLanguage: handleSetStoredSelectedLanguage
  }
}
