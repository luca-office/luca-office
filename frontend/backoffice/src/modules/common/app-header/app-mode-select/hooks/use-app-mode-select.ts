import {useCallback, useState} from "react"
import {useDispatch, useSelector} from "react-redux"
import {useCheckLogin} from "shared/graphql/hooks"
import {UserAccountForLogin} from "shared/models"
import {Option} from "shared/utils"
import {AppMode} from "../../../../../enums"
import {navigateToRouteAction} from "../../../../../redux/actions/navigation-action"
import {updateAppMode} from "../../../../../redux/actions/ui/common-ui-action"
import {AppState} from "../../../../../redux/state/app-state"
import {Route} from "../../../../../routes"

export interface UseAppModeSelectHook {
  readonly appMode: AppMode
  readonly changeAppMode: (mode: AppMode) => void
  readonly isMenuVisible: boolean
  readonly toggleMenu: () => void
  readonly account: Option<UserAccountForLogin>
}

export const useAppModeSelect = (): UseAppModeSelectHook => {
  const [isMenuVisible, toggleMenuVisibility] = useState(false)
  const appMode = useSelector<AppState, AppMode>(state => state.ui.common.appMode)

  const {account} = useCheckLogin()
  const dispatch = useDispatch()

  const changeAppMode = useCallback((appMode: AppMode) => {
    dispatch(updateAppMode(appMode))

    switch (appMode) {
      case AppMode.MANAGER:
        dispatch(navigateToRouteAction(Route.Projects))
        break
      case AppMode.USERMANAGEMENT:
        dispatch(navigateToRouteAction(Route.UserManagement))
        break
      case AppMode.RATING:
        dispatch(navigateToRouteAction(Route.RaterRatingOverview))
        break
      case AppMode.EDITOR:
      default:
        dispatch(navigateToRouteAction(Route.Scenarios))
    }
    toggleMenuVisibility(false)
  }, [])

  const toggleMenu = () => {
    toggleMenuVisibility(!isMenuVisible)
  }

  return {
    appMode,
    changeAppMode,
    isMenuVisible,
    toggleMenu,
    account
  }
}
