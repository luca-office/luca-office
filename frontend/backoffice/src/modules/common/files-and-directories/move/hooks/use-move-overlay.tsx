import {Dispatch, SetStateAction, useState} from "react"
import {FileType} from "shared/enums"
import {useUpdateFile} from "shared/graphql/hooks"
import {Directory, File} from "shared/models"
import {findPathToRoot, Option} from "shared/utils"
import {useUpdateDirectory} from "../../../../../graphql/hooks/mutations/directory/update-directory"

export interface UseMoveOverlay {
  readonly selectedTargetDirectory: Option<UUID>
  readonly setSelectedTargetDirectory: Dispatch<SetStateAction<Option<string>>>
  readonly changeDirectory: () => void
}

// Use this if you need an id for root and can not use null
export const ROOT = "root_identifier"

export const useMoveOverlay = (
  directoryOption: Option<Directory>,
  fileOption: Option<File>,
  onFilesSuccessfullyMoved: (targetDirectoryIdWithParents: UUID[]) => void,
  directories: Directory[],
  files: File[]
): UseMoveOverlay => {
  const [selectedTargetDirectory, setSelectedTargetDirectory] = useState<Option<UUID>>(Option.none())

  const {updateDirectory} = useUpdateDirectory()
  const {updateFile} = useUpdateFile()

  const handleChangeDirectory = () => {
    directoryOption.forEach(directory =>
      selectedTargetDirectory.forEach(targetDirectory => {
        const parentDirectoryId = targetDirectory === ROOT ? null : targetDirectory

        const parentDirectoryIdWithParents = getTargetDirectoryIdsWithParents(targetDirectory, directories, files)
        updateDirectory(directory.id, {name: directory.name, parentDirectoryId}).then(() =>
          onFilesSuccessfullyMoved(parentDirectoryIdWithParents)
        )
      })
    )

    fileOption.forEach(file =>
      selectedTargetDirectory.forEach(targetDirectory => {
        const parentDirectoryIdWithParents = getTargetDirectoryIdsWithParents(targetDirectory, directories, files)
        updateFile(file.id, {
          binaryFileId: file.fileType === FileType.Media ? file.binaryFileId : null,
          name: file.name,
          relevance: file.relevance,
          tags: file.tags,
          directoryId: targetDirectory
        }).then(() => onFilesSuccessfullyMoved(parentDirectoryIdWithParents))
      })
    )
  }

  return {
    selectedTargetDirectory,
    setSelectedTargetDirectory,
    changeDirectory: handleChangeDirectory
  }
}

const getTargetDirectoryIdsWithParents = (targetDirectoryId: UUID, directories: Directory[], files: File[]) => {
  const pathToRoot = findPathToRoot(targetDirectoryId, directories, files)
  return [...pathToRoot, targetDirectoryId]
}
