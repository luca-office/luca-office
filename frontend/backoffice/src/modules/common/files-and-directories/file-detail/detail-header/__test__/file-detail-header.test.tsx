import {MockedProvider} from "@apollo/client/testing"
import {mount} from "enzyme"
import * as React from "react"
import {create} from "react-test-renderer"
import {CardHeader, DeleteOrArchiveEntityButton, Heading, Icon} from "shared/components"
import {IconName} from "shared/enums"
import {useDeleteFileFromScenario} from "shared/graphql/hooks"
import {DetailHeader, FileDetailHeaderProps} from "../detail-header"

const defaultProps: FileDetailHeaderProps = {
  deleteHook: useDeleteFileFromScenario,
  entityId: "pofjsdf-afpaojfs",
  icon: IconName.Image,
  title: "file.png",
  disabled: false
}

const useComponent = (props?: Partial<FileDetailHeaderProps>) => (
  <MockedProvider>
    <DetailHeader {...defaultProps} {...props} />
  </MockedProvider>
)

describe("fileDetailHeader", () => {
  it("renders correctly", () => {
    const component = useComponent()
    const tree = create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("has correct structure", () => {
    const component = mount(useComponent())

    expect(component.find(CardHeader)).toHaveLength(1)
    expect(component.find(Icon)).toHaveLength(2)
    expect(component.find(Heading)).toHaveLength(1)
    expect(component.find(DeleteOrArchiveEntityButton)).toHaveLength(1)
  })
})
