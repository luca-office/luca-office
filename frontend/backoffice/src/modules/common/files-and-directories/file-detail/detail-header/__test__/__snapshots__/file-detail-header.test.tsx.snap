// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fileDetailHeader renders correctly 1`] = `
<div
  className="card-header css-13nbzsf-CardHeader"
>
  <div
    className="css-17beenm-DetailHeader"
  >
    <div
      className="css-1lrzij5-Icon"
      dangerouslySetInnerHTML={
        {
          "__html": {},
        }
      }
    />
    <div
      className="css-1i6jmor-Heading"
    >
      file.png
    </div>
  </div>
  <div
    className="css-ntx7po-DetailHeader"
  >
    <div
      className="css-0"
    >
      <button
        className="css-hnekdl-Button"
        disabled={false}
        onClick={[Function]}
      >
        <div
          className="css-95j2hm-Button"
        >
          <div
            className="css-mj8vo1-Icon"
            dangerouslySetInnerHTML={
              {
                "__html": {},
              }
            }
          />
        </div>
      </button>
    </div>
  </div>
</div>
`;
